ECP-WarpX/WarpX
# Set up CTest

Work-in-progress PR to set up CTest. Apply the same logic as in ImpactX and adjust/advance for WarpX.

- Example of current input files organization:
  - `inputs_base_2d`
    (base input file for shared input parameters named as `inputs_base_<1d;2d;3d;rz>`)
  - `inputs_test_2d_langmuir_multi_nodal`,
    `inputs_test_2d_langmuir_multi_psatd`,
    ...
    (test input files for test-specific input parameters named as `inputs_test_<1d;2d;3d;rz>_...`, read base input file through the `FILE` parameter)
  - test names must start with `test_<1d;2d;3d;rz>_...` to guarantee consistency and make maintenance easier

- New Azure CI matrix elements (list coming soon).

- Run all MPI tests with 2 MPI processes and 1 OpenMP thread. Run single-process tests without MPI.

To-do:
- [x] `WarpX_TEST_CLEANUP` to remove test directories
- [x] `WarpX_TEST_FPETRAP` for runtime fpe-trapping
- [x] Resolve `FIXME` comments or redirect to new issues
- [x] Avoid copying inputs files https://github.com/AMReX-Codes/amrex/pull/4126
- [x] Enforce new naming convention through CI style check
- [x] Remove obsolete CI style check scripts
- [x] Remove obsolete regression testing scripts: #5220
- [x] Remove files after each test: new build flag `WarpX_CI`
- [x] Make sure we don't miss any build variant:
  - [x] single-process Python
  - [x] embedded boundaries
  - single- & mixed-precision (new issue #5205)
- [x] Add missing checksums (new issue #5206):
  - `dirichlet_bc`
  - `dirichlet_bc_picmi`
  - `field_probe`
  - `embedded_boundary_picmi`
  - `particle_absorption`
  - `particle_reflection_picmi`
  - [x] `id_cpu_read_picmi`
  - `runtime_components_picmi`
  - `laser_acceleration_rz_opmd`
- [x] Support restart tests
  - [x] Separate input file based on original input file (read through the `FILE` parameter)
  - [x] Call default regression analysis against original checksum file, with small tolerance
  - [x] Support restart from PICMI (can we set `FILE` through PICMI?)
  - [x] Update `analysis_default_restart.py`
- [x] Support custom run commands (e.g., custom PICMI scripts using lists vs strings)
- Add stricter dependency between `.analysis` test and `.run` test:
  run `.analysis` test only if `.run` test passed (now runs if `.run` test ran)
- Update test README files (e.g., Ohm solver tests)
- [ ] Skip pyAMReX tests directly from CMakeLists.txt
